Feature/launcher

m3d opened the pull request:
This PR is about introduction of "launcher" which would be common for various robot applications. `navpat.py` is used as example. For now the functionality is "brutally" split between two files. As expected there are several "unknowns"/questions:

1) should viewer (option --view) be handled by launcher? (I suppose `yes`)
(+) other applications will automatically have viewer available without need of any code
(-) there have to be generic and specific viewer extensions, i.e. "detected cones" which are specific for `navpat.py`

2) should launcher attach processor, i.e. external processing? I suppose that `no`, because this code is specific to application. Unfortunately then it has to share `metalog` to be able to call `attach_processor`.

3) who should handle `EmergencyStopException`? This is exception thrown by robot when any of the STOP buttons is pressed. I suppose that this should be handled by launcher, but then we have to pass our main method to it, so that launcher wraps it into try/except block.

4) who should clean-up and call optional viewer? Now it is `navpat.ph` but it should be rather launcher ... but then it should be maybe `class Launcher` which would remember results of parsing? (in particular `--view`)

5) should have the application access to `config`? I suppose that `yes` (I expect that `navpat` will store there mowing pattern) but maybe only part related to the application?

zwn commented:
> This PR is about introduction of "launcher" which would be common for various robot applications.

It is called `launcher` but it seems to be specific to john deere robot. If that is so, can we make it part of `johndeere.py`? As a generic "launcher" deserving the name `launcher.py` I would see a tool common to all robots (the one that reads names of modules to load/start from the config).

> should launcher attach processor, i.e. external processing?

General launcher yes, this launcher no. If the config is used to name the modules to start, one of them can be "processor" (that thing that runs in a separate proces).

> I suppose that no, because this code is specific to application. Unfortunately then it has to share metalog to be able to call attach_processor.

I think there should be an API on the robot class. I mean, not just member variables that you can set, but real method you can call. I think the logging stuff should never escape outside from this robot class. Then `attach_processor(robot, metalog, callback)` would be `robot.attach_processor(callback)`. So far I think we have

```python
robot.update()
robot.attach_processor(id, callback) # callback called for data with id in a different process
robot.attach_extension(id, callback) # callback called within the main thread during update
```

Callback returns `(id, data)` that is handled during `update()` (next one for processor, current one for extension).

> who should handle EmergencyStopException? This is exception thrown by robot when any of the STOP buttons is pressed. I suppose that this should be handled by launcher, but then we have to pass our main method to it, so that launcher wraps it into try/except block.

I think the way to handle this is to make the `robot` be a context manager:

```python
with parse_and_launch(sys.argv) as robot:
    navigate_pattern(robot)
```

Here when `robot` goes out of scope the exit method is called where cleanup can be done (like stopping the motor). That is the python way to do C++ destructors.

I am not sure yet about the viewer stuff. It is somewhat confusing to me right now. Can we somehow "ignore it" and maybe add it again later when we have the basic sorted out?

For me the basics are the API on the robot class including the context manager with no (meta)log visible anywhere outside of the implementation.

m3d commented:
I like the idea of context manager (added fixup) :-). It partially solves also the `--view` option. I also agree that the application should not have access to metalog, and proposed solution with `robot.attach_processor(image_callback)` would be nice, but I just do not know how to implement it now (without ugly dynamic extension of robot or adding metalog to robot). Suggestions welcome.

zwn commented:
Why is adding metalog to the robot object ugly? I don't have any new suggestions. Just the old ones:
- hide metalog inside robot.
- don't use `helper` in `navpat`
- don't name it `launcher` since it is specific to john deere robot
- decide on an API for the robot object that does not bleed the implementation to the outside (like for example canproxy)

zwn commented:
Notice that all functions inside `helper` are actually methods on `robot`. Just do `s/robot/self/` :wink:

zwn commented:
I see the robot class as a Hardware Abstraction Layer. For the John Deere tractor it has CAN, Camera, Laser and other classes/devices/extensions but it translates this info into a common API. Once we have the other tractors working, all we'll have to do is to implement the same API.

m3d commented:
Well, I would keep `launcher` for robots launching, although we have only one type supported at the moment. I would not add more functionality until we have two (or three?) robots ready to integrate. In particular I do not agree with "don't name it launcher since it is specific to john deere robot".

Regarding "Hardware Abstraction Layer" ... so should the robot be abstract or not? In particular I do not like to move all the hardware/sensor dependent code from `helper.py`. The idea of sources and extension is probably common so that could be moved to `Robot`.

zwn commented:
All I am asking is way for "application developer" not to deal with the innards of the implementation. If you do not like what I am suggesting, suggest something else otherwise we will not move very far.

m3d commented:
Rebased on master with fixes from `feature/test170808` branch. Changed description of `launcher.py`, which has still value to simplify "launching" of other applications. Provided `helloworld.py` as sample application, but changes are expected with introduction of new robot API.

I would postpone reading configuration from zip file to another PR.
